File: compiler/bench/leapyear.fs
\ expected: 485
\ Leap year benchmark - count leap years 1-2000, run 1000 times

: leap-year? ( y -- flag )
  dup 400 mod 0= if drop -1 exit then
  dup 100 mod 0= if drop 0 exit then
  4 mod 0= ;

: count-leap ( limit -- count )
  0 swap 1+ 1 do
    i leap-year? if 1+ then
  loop ;

: main ( -- )
  0
  1000 0 do
    drop 2000 count-leap
  loop
  . cr ;

File: compiler/bench/leapyear.c
// expected: 485
// Leap year benchmark - count leap years 1-2000, run 1000 times

#include <stdio.h>

int leap_year(int y) {
    if (y % 400 == 0) return 1;
    if (y % 100 == 0) return 0;
    return y % 4 == 0;
}

int count_leap(int limit) {
    int count = 0;
    for (int i = 1; i <= limit; i++) {
        if (leap_year(i)) count++;
    }
    return count;
}

int main(void) {
    int result = 0;
    for (int i = 0; i < 1000; i++) {
        result = count_leap(2000);
    }
    printf("%d\n", result);
    return 0;
}
